_posts/2024-08-01-search-standalone-mapper.adoc

```diff
@@ -0,0 +1,295 @@
+    }
+}
+----
+1. Get an iterator from the finite data stream of guides.
+2. Read the next batch of the guides from the iterator. We are using the batch size limit
+that we will retrieve from the mass-loading options
+and checking the iterator to see if there are any more entities to pull.
+====
+
+Now with having the way of reading the entities in batches from the stream
+and knowing how to pass it to the mass indexer, implementing the identifier loader
+can be as easy as:
+
+[source, java]
+====
+----
+@Override
+public MassIdentifierLoader createIdentifierLoader(LoadingTypeGroup<Guide> includedTypes,
+        MassIdentifierSink<Guide> sink, MassLoadingOptions options) {
+    GuideLoadingContext context = options.context(GuideLoadingContext.class); // <1>
+    return new MassIdentifierLoader() {
+        @Override
+        public void close() {
+            // nothing to do
+        }
+
+        @Override
+        public long totalCount() {
+            return 0; // <2>
+        }
+
+        @Override
+        public void loadNext() throws InterruptedException {
+            List<Guide> batch = context.nextBatch(options.batchSize()); // <3>
+            if (batch.isEmpty()) {
+                sink.complete();  // <4>
+            } else {
+                sink.accept(batch); // <5>
+            }
+        }
+    };
+}
+----
+1. Retrieve the guide loading context that is expected to be passed to the mass indexer.
+(e.g. `.context(GuideLoadingContext.class, guideLoadingContext)`)
+2. We do not know how many guides there will be until we finish reading all the files
+and completing the indexing, so we'll just pass `0` here.
++
+NOTE: This is one of the areas that we plan to improve;
+see https://hibernate.atlassian.net/browse/HSEARCH-5180[one of the improvements we are currently working on].
+3. Request the next batch of guides. `options.batchSize()` will provide us with the value configured
+for the current mass indexer.
+4. If the batch is empty, it means that the stream iterator has no more guides to return.
+Hence, we can notify the mass indexing sink that no more items will be provided by calling `.complete()`.
+5. If there are any guides in the loaded batch, we'll pass them to the sink to be processed.
+====
+
+
+To sum things up, here is a summary of which steps to take to index unknown number of search entities from a datasource
+from which we would want to read just once and do not do any entity lookups by id:
+
+1. Start by creating a loader binder and let Hibernate Search know about it via the `@SearchEntity` annotation.
+2. Implement a mass loading strategy (`MassLoadingStrategy`) that includes:
+  - An identifier loader that does all the heavy lifting and actually constructs entire entities.
+  - An entity loader that simply passes through the entities loaded by the identifier loader to the indexing sink.
+3. Supply any helpful services, resources, helpers, etc., that are used to load the data through the mass indexer context.
```

Comment:
To avoid misunderstanding the end of the sentence ("load data through the mass indexer context") as "use the mass indexer context to load data":

```suggestion
3. Supply through the mass indexer context any helpful services, resources, helpers, etc., that are used to load the data.
```